Configure severity level of .editorconfig rule from editor with Additional Files

allisonchou opened the pull request:
Added support to configure the severity level of an .editorconfig rule directly through the editor. If a user does not currently have an .editorconfig file at the project level, one will be generated for them.

Example if user already has pre-existing .editorconfig file (marked as Additional File):
![configexample1](https://user-images.githubusercontent.com/16968319/44180220-14e46200-a0b0-11e8-8beb-1288ce5f7962.png)

Example if user does not have a pre-existing .editorconfig file:
![configexample2](https://user-images.githubusercontent.com/16968319/44180224-17df5280-a0b0-11e8-9fe8-79bd4ab2ffd4.png)

**Note:** This feature is currently in its very early iterations and is expected to see large changes in the future.
Some things of note:
1. Due to technical limitations, this feature currently only works if an .editorconfig file is marked as an Additional File at the project-level. If there is an existing .editorconfig file at the project-level but it is not marked as an Additional File, then the file will not be recognized. In future iterations of this feature, we likely want to consider a solution such as automatically marking an .editorconfig file as an Additional File upon the file being added to the project/solution, or allowing an .editorconfig file to recognized as a regular document and getting rid of the Additional Document workaround altogether.

2. Adding an .editorconfig and marking it as an Additional File is currently not supported by Roslyn (see issue [#4655](https://github.com/dotnet/roslyn/issues/4655)). Therefore, while we can currently add the .editorconfig file for the user, the user must mark it as an Additional File manually.

CyrusNajmabadi commented on src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs:
```diff
@@ -131,12 +133,32 @@ public CSharpAddAccessibilityModifiersDiagnosticAnalyzer()
+                }
+                else
+                {
+                    properties["OptionCurrent"] = option.Value.ToString().ToLowerInvariant();
```
so.  in general, i'm pretty concerned about this sort of duplication happening all over the place.  It's not at all clear, IMO, for features to know tey need to do this sort of this.  It's also not clear why this wouldn't just be specified on the *option* itself.  Then the option could just be used to create all this data.

CyrusNajmabadi replied:
still have this concern.

CyrusNajmabadi commented on src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs:
```diff
@@ -131,12 +133,32 @@ public CSharpAddAccessibilityModifiersDiagnosticAnalyzer()
             var additionalLocations = ImmutableArray.Create(member.GetLocation());
+
+            var properties = ImmutableDictionary.CreateBuilder<string, string>();
+            var ruleName = CodeStyleOptions.RequireAccessibilityModifiers.StorageLocations.OfType<EditorConfigStorageLocation<CodeStyleOption<AccessibilityModifiersRequired>>>().FirstOrDefault();
```
I really dislike all of this.  It makes things so leaky.  

Ths is similar to my issue with how we did codecleanup in the first place.  Instead of keeping concerns separated, we made it so that different systems have a deep and leaky knowledge of how other parts of the ysstem work.  This is extremely brittle and difficult to maintain.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Suppression/CSharpConfigureSeverityLevel.cs:
```diff
@@ -0,0 +1,39 @@
+            { IDEDiagnosticIds.UseExpressionBodyForOperatorsDiagnosticId, CSharpCodeStyleOptions.PreferExpressionBodiedOperators },
+            { IDEDiagnosticIds.UseExpressionBodyForPropertiesDiagnosticId, CSharpCodeStyleOptions.PreferExpressionBodiedProperties },
+            { IDEDiagnosticIds.UseExpressionBodyForIndexersDiagnosticId, CSharpCodeStyleOptions.PreferExpressionBodiedIndexers },
+            { IDEDiagnosticIds.UseExpressionBodyForAccessorsDiagnosticId, CSharpCodeStyleOptions.PreferExpressionBodiedAccessors }
```
This continues the tendancy for us to have tight coupling between features scattered all over the codebase.  I very much do not like this.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,283 @@
+            { IDEDiagnosticIds.UseDeconstructionDiagnosticId, CodeStyleOptions.PreferDeconstructedVariableDeclaration },
+            { IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId, CodeStyleOptions.PreferReadonly },
+            { IDEDiagnosticIds.UseConditionalExpressionForAssignmentDiagnosticId, CodeStyleOptions.PreferConditionalExpressionOverAssignment },
+            { IDEDiagnosticIds.UseConditionalExpressionForReturnDiagnosticId, CodeStyleOptions.PreferConditionalExpressionOverReturn }
```
More tight coupling.  How would anyone know they would ever need to update these areas of the code when changing any sort of feature/analyzer/fixer/etc.?

CyrusNajmabadi commented:
As an example.  Imagine we wanted to light stuff up for VB or C# specific features.  Could we do that without adding specific code that had to know about those features' IDs/Options themselves?  If not, that's a problem, because we're effectively saying: the only way to Cleanup/Configure something is for us to hardcode knowledge about it.  That violates core concepts we have with Analyzers/Fixers/Refactorings in that they need to be something that can just be added to the system independently without anything else having to know or care about them.

mavasani commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,283 @@
+        public ConfigureSeverityLevelCodeAction(
+            Diagnostic diagnostic,
+            ImmutableArray<CodeAction> nestedActions)
+            : base("Configure severity level via editorconfig file",
```
This should be a localizable resource. Can you create a resource string for it?

mavasani commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,283 @@
+        {
+            var solution = project.Solution;
+            TextDocument editorconfig = null;
+            FindOrGenerateEditorconfig();
```
Change signature to return the editorconfig?

mavasani commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,283 @@
+            string mostRecentHeader = null;
+
+            var name = string.Empty;
+            FindRuleName();
```
Change signature to `string FindRuleName()` to return the name?

mavasani replied:
Nit: Return `null` if cannot find it.

mavasani commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,283 @@
+            TextDocument editorconfig = null;
+            FindOrGenerateEditorconfig();
+
+            editorconfig.TryGetText(out var text);
```
Unused?

mavasani commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,283 @@
+            FindOrGenerateEditorconfig();
+
+            editorconfig.TryGetText(out var text);
+            var result = editorconfig.GetTextAsync().Result;
```
Can you instead make this an async method `internal static async Task<Solution> ConfigureEditorConfigAsync(...)` and use `var result = await editorconfig.GetTextAsync().ConfigureAwait(false)` here?

mavasani commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,283 @@
+
+                    var rulePattern = new Regex(@"([\w ]+)=([\w ]+):([\w ]+)");
+                    var headerPattern = new Regex(@"\[\*([^#\[]*)\]");
+                    if (rulePattern.IsMatch(curLineText))
```
You should probably add a few unit tests where the header or rule entry has typo.

allisonchou commented:
Pushed in a few more refactorings. Any feedback is greatly appreciated!

CyrusNajmabadi commented:
Some questions about the feature overall:

![image](https://user-images.githubusercontent.com/4564579/44495592-597a7b00-a625-11e8-81ac-f1f58c9cc2b3.png)

Overall i have a general concern about this bloating up the context menu too much.  Is it possible you could have a few suggestions here and each suggestion will have a corresponding "configure severity" option shown along with it?  Should we instead potentially expose this on hte preview dialog like we do for 'Fix all'.  i.e. here:

![image](https://user-images.githubusercontent.com/4564579/44495643-ac543280-a625-11e8-80ff-196b4770f74e.png)

Can we add the "configure severity as a drop down *here* instead?  That way hte main list stays uncluttered, but config about a specific fix is all centralized?

CyrusNajmabadi commented:
Question about:

![image](https://user-images.githubusercontent.com/4564579/44495682-d3aaff80-a625-11e8-9cf0-825567f02ca6.png)

Do we need "Refactoring Only" as that's another severity we support?

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:
```diff
@@ -489,6 +489,13 @@ private static async Task VerifyAgainstWorkspaceDefinitionAsync(string expectedT
                         var expectedRoot = await expectedDocument.GetSyntaxRootAsync();
                         Assert.Equal(expectedRoot.ToFullString(), root.ToFullString());
                     }
+                    foreach (var additionalDoc in project.AdditionalDocuments)
```
nit.  blank line above this.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+            Diagnostic diagnostic,
+            ImmutableArray<CodeAction> nestedActions)
+            : base(FeaturesResources.Configure_severity_level_via_editorconfig_file,
+            ImmutableArray<CodeAction>.CastUp(nestedActions), isInlinable: false)
```
this CastUp seems unnecessary.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+        {
+            var solution = project.Solution;
+            var editorconfig = FindOrGenerateEditorconfig();
+            var result = await editorconfig.GetTextAsync().ConfigureAwait(false);
```
pass cancellationToken to GetTextAsync.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+            Diagnostic diagnostic,
+            Project project,
+            Dictionary<string, Option<CodeStyleOption<bool>>> languageSpecificOptions,
+            Dictionary<string, Option<CodeStyleOption<ExpressionBodyPreference>>> expressionOptions,
```
this seems like a layering violation.  This is a generic system for configuring editorconfig.  Why does it have any knowledge of specific enums like ExpressionBodyPreference?  Would we expect that every time we added a new optoin, we'd need to come fix up this code to be aware of it?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+                // First we check if the rule already exists in the file; if it does, we replace it
+                if (CheckIfRuleExistsAndReplaceInFile())
+                {
+                    return solution;
```
if hte rule exists... why would we offer the feature in the first place?   That concerns be.  That means that even if the user has a rule for this feature, we'll be bloating the lightbulb menu for all future invocations?  I don't think that's an ok user experience.

CyrusNajmabadi replied:
I don't understand this pattern.  All codepaths end up returning 'solution'.  so what actually causes anything to change?

allisonchou replied:
1. Based on customer interviews Kasey and I conducted, some customers indicated that they wanted to change the severity level of rules depending on what stage of the code development process they were in. So for example, in order for people to pay greater attention to fixing a diagnostic, they may now want a rule that was previously configured as a warning to now be configured as an error.

2. The change is caused by calling solution.WithAdditionalDocument(…) and ultimately returning the solution returned from the method.

CyrusNajmabadi replied:
sorry, my point wasn't clear.  What i was trying to say was: this makes it look as if you're returning the 'solution' *untouched*.  That's because you actually mutate it in non-local ways because of your use of local functions.  That makes the code actually look like it's doing something else.

CyrusNajmabadi replied:
See: https://github.com/dotnet/roslyn/pull/29332#discussion_r212139780

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+                    {
+                        doc = curDoc;
+                    }
+                }
```
`var doc = project.AdditionalDocuments.FirstOrDefault(d => d.Name...)`.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+# will not correctly take effect until additional steps are taken to enable it. See the
+# following page for additional information:
+#
+# https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Enable%20Editorconfig%20Configuration.md";
```
needs loc.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+# following page for additional information:
+#
+# https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Enable%20Editorconfig%20Configuration.md";
+                    solution = solution.AddAdditionalDocument(id, ".editorconfig", editorconfigDefaultFileContent);
```
oh... i really dislike this.  it's not clear at all from reading the code in the main method that you will be overwriting this local in a nested function.  Please refactor this code to not do that :)

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+                    solution = solution.AddAdditionalDocument(id, ".editorconfig", editorconfigDefaultFileContent);
+                    doc = solution.GetAdditionalDocument(id);
+                }
+                return doc;
```
newline after }

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs:
```diff
@@ -0,0 +1,284 @@
+                if (diagnosticToEditorConfigDotNet.ContainsKey(diagnostic.Id))
+                {
+                    diagnosticToEditorConfigDotNet.TryGetValue(diagnostic.Id, out var value);
+                    var storageLocation = value.StorageLocations.OfType<EditorConfigStorageLocation<CodeStyleOption<bool>>>().FirstOrDefault();
```
i've seen this pattern all over hte place.  It's not good.  we are leaking out inner details of the options.

CyrusNajmabadi commented:
The concerns i brought up here still apply: https://github.com/dotnet/roslyn/issues/29345

Namely that we are very *significantly* adding severe technical debt here with this approach. 

To put it in perspective:  If i were to create a new IDE feature i would have *no idea* what subset of the IDE i would now have to update for it to properly work with all these new subsystems we're building.  It's already hard enough to figure out everythgin that needs updating when you add an option+analyzer+refactoring+fixer.  It's now become even *harder*.  You have to know exactly what needs updating in code cleanup.  You have to know exactly what needs updating in .editorconfig generation.  And now you have to know exactly what needs updating in this suppression area.

This is not healthy.  It not only deeply isolates IDE features as the only features "blessed" play in these spaces, but it also means people have to deeply understand the interrelationship of all of these subsystems in order to do any work at all.

I know it is not palatable to hear, but i strongly do not believe it's acceptable for this to go in in the current form *unless* it is into a feature branch.  I definitely applaud the work here and i think there is huge value in bringing these experiences to users.  However, that work must be done in a manner that preserves hte health of hte codebase and does not introduce large amounts of coupling, brittleness, and silo'ed knowledge, like what we see here.

allisonchou commented:
Thank you for the helpful feedback, Cyrus! I realize that this feature has a long way to go, most especially due to architectural limitations and open-ended design questions. The hard-coding and tight coupling is something that will most definitely have to be addressed before the feature is considered for release. This current iteration will only go into the .editorconfig feature branch and a design meeting addressing your concerns will likely occur in the future. I appreciate the feedback!

mavasani commented:
@allisonchou @CyrusNajmabadi - I have a [PR](https://github.com/dotnet/roslyn/pull/35691) out to revive this work with .editorconfig as an `AnalyzerConfigDocument`, so we no longer need .editorconfig file to be an additional document. I have also removed the hard-coding that was Cyrus' main concern and moved the mapping of diagnostic ID to code style option into a [service](https://github.com/dotnet/roslyn/pull/35691/files#diff-336be816aefb8fcb32f99f49b0fff109R12) with [unit tests](https://github.com/dotnet/roslyn/pull/35691/files#diff-9cde800c8b09a533f1e382934e0f7137R16), which will break if anyone adds a new IDE diagnostic ID but does not handle this new ID in the added service [here](https://github.com/dotnet/roslyn/pull/35691/files#diff-41b1246e47f377a358e64aad518db68cR107).
